main.cpp

```diff
@@ -1,109 +1,228 @@
-        printf("Error! socket.open() returned: %d\n", result);
-        goto DISCONNECT;
+        printf("%s address is %s\r\n", hostname, (address.get_ip_address() ? address.get_ip_address() : "None") );
+
+        return true;
     }
 
-    // Get the host address
-    printf("\nResolve hostname %s\n", hostname);
-    result = net->gethostbyname(hostname, &a);
-    if (result != 0) {
-        printf("Error! gethostbyname(%s) returned: %d\n", hostname, result);
-        net->disconnect();
-        exit(-1);
+    bool send_http_request()
+    {
+        /* loop until whole request sent */
+        const char buffer[] = "GET / HTTP/1.1\nHost: ifconfig.io\nConnection: close\n\n";
+
+        nsapi_size_t bytes_to_send = strlen(buffer);
+
+        while (bytes_to_send) {
+            nsapi_size_or_error_t result = socket.send(buffer + result, bytes_to_send);
+            if (result < 0) {
+                printf("Error! socket.send() returned: %d\r\n", result);
+                return false;
+            }
+
+            bytes_to_send -= result;
+        }
+
+        printf("\r\nsent %d bytes: \r\n%s", strlen(buffer), buffer);
+
+        return true;
     }
-    printf("%s address is %s\n", hostname, (a.get_ip_address() ? a.get_ip_address() : "None") );
-
-    a.set_port(80);
-    result = socket.connect(a);
-    if (result != 0) {
-        printf("Error! socket.connect() returned: %d\n", result);
-        net->disconnect();
-        exit(-2);
+
+    bool receive_http_response()
+    {
+        char buffer[MAX_MESSAGE_RECEIVED_LENGTH];
+        int remaining_bytes = MAX_MESSAGE_RECEIVED_LENGTH;
+        int received_bytes = 0;
+
+        /* loop until there is nothing received or we've ran out of buffer space */
+        while (remaining_bytes > 0) {
+            nsapi_size_or_error_t result = socket.recv(buffer + received_bytes, remaining_bytes);
+            if (result < 0) {
+                printf("Error! socket.recv() returned: %d\r\n", result);
+                return false;
+            }
+
+            received_bytes += result;
+            remaining_bytes -= result;
+        }
+
+        /* the message is likely larger but we only want the HTTP response code */
+
+        printf("received %d bytes:\r\n%.*s\r\n\r\n", received_bytes, strstr(buffer, "\n") - buffer, buffer);
+
+        return true;
     }
 
-    // Loop until whole request sent
-    while (size) {
-        result = socket.send(sbuffer+result, size);
-        if (result < 0) {
-            printf("Error! socket.send() returned: %d\n", result);
-            goto DISCONNECT;
+    void wifi_scan()
+    {
+        WiFiInterface *wifi = _net->wifiInterface();
+
+        WiFiAccessPoint ap[MAX_NUMBER_OF_ACCESS_POINTS];
+
+        /* scan call returns number of access points found */
+        int result = wifi->scan(ap, MAX_NUMBER_OF_ACCESS_POINTS);
+
+        if (result <= 0) {
+            printf("WiFiInterface::scan() failed with return value: %d\r\n", result);
+            return;
```

Comment:
Is it expected that if the scan fail it tries to connect when the function returns ?

Reply:
the scan can fail but connection can still succeed (although probably unlikely in practice)